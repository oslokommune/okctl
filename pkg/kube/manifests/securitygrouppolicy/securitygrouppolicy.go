// Package securitygrouppolicy provides a convenient way of
// interacting with the SecurityGroupPolicy CRD
package securitygrouppolicy

import (
	"context"
	"fmt"

	v1beta1client "github.com/oslokommune/okctl/pkg/kube/securitygrouppolicy/clientset/v1beta1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

	"github.com/oslokommune/okctl/pkg/kube/securitygrouppolicy/api/types/v1beta1"
	restclient "k8s.io/client-go/rest"
)

// SecurityGroupPolicy contains the required state
type SecurityGroupPolicy struct {
	Name      string
	Namespace string
	Manifest  *v1beta1.SecurityGroupPolicy

	Config *restclient.Config
	Ctx    context.Context
}

// New returns an initialised client for interacting with security group policies
func New(name, namespace string, manifest *v1beta1.SecurityGroupPolicy, config *restclient.Config) *SecurityGroupPolicy {
	return &SecurityGroupPolicy{
		Name:      name,
		Namespace: namespace,
		Manifest:  manifest,
		Config:    config,
		Ctx:       context.Background(),
	}
}

// Create the security group policy if it doesn't exists
func (s *SecurityGroupPolicy) Create() (*v1beta1.SecurityGroupPolicy, error) {
	client, err := v1beta1client.NewForConfig(s.Config)
	if err != nil {
		return nil, fmt.Errorf("building client from config: %w", err)
	}

	policies, err := client.SecurityGroupPolicy(s.Namespace).List(s.Ctx, metav1.ListOptions{})
	if err != nil {
		return nil, fmt.Errorf("listing security group policies; %w", err)
	}

	for _, policy := range policies.Items {
		if policy.Name == s.Name {
			return &policy, nil
		}
	}

	p, err := client.SecurityGroupPolicy(s.Namespace).Create(s.Ctx, s.Manifest)
	if err != nil {
		return nil, fmt.Errorf("applying security group policy: %w", err)
	}

	return p, nil
}

// Delete the security group policy
func (s *SecurityGroupPolicy) Delete() error {
	client, err := v1beta1client.NewForConfig(s.Config)
	if err != nil {
		return err
	}

	policy := metav1.DeletePropagationForeground

	return client.SecurityGroupPolicy(s.Namespace).Delete(s.Ctx, s.Name, metav1.DeleteOptions{
		PropagationPolicy: &policy,
	})
}

// Manifest returns a SecurityGroupPolicy manifest
func Manifest(name, namespace string, matchLabels map[string]string, securityGroups []string) *v1beta1.SecurityGroupPolicy {
	return &v1beta1.SecurityGroupPolicy{
		TypeMeta: metav1.TypeMeta{
			Kind:       "SecurityGroupPolicy",
			APIVersion: "vpcresources.k8s.aws/v1beta1",
		},
		ObjectMeta: metav1.ObjectMeta{
			Name:      name,
			Namespace: namespace,
		},
		Spec: v1beta1.SecurityGroupPolicySpec{
			PodSelector: v1beta1.SecurityGroupPolicyPodSelector{
				MatchLabels: matchLabels,
			},
			SecurityGroups: v1beta1.SecurityGroupPolicySecurityGroups{
				GroupIDs: securityGroups,
			},
		},
	}
}
