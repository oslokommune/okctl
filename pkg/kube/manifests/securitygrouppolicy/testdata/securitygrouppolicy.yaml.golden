apiVersion: vpcresources.k8s.aws/v1beta1
kind: SecurityGroupPolicy
metadata:
  creationTimestamp: null
  name: name
  namespace: default
spec:
  podSelector:
    matchLabels:
      app: psqlclient
  securityGroups:
    groupIds:
    - sg-a90eug3FAKE
